File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Related_2/_155_Minimum_Depth_of_Binary_Tree_Easy.java
package J_3_Binary_Tree_Divide_Conquer.Related_2;
import java.util.*;
/**
155
Minimum Depth of Binary Tree

 * Created by tianhuizhu on 6/28/17.
 */
public class _155_Minimum_Depth_of_Binary_Tree_Easy {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public class Solution {
        public int minDepth(TreeNode root) {
            if (root == null) {
                return 0;
            }
            return getMin(root);
        }

        public int getMin(TreeNode root){
            if (root == null) {
                return Integer.MAX_VALUE;
            }

            if (root.left == null && root.right == null) {
                return 1;
            }

            return Math.min(getMin(root.left), getMin(root.right)) + 1;
        }
    }
}

File: JS9Ch/src/J_3_Binary_Tree_Divide_Conquer/Related_2/python_cpp/_155_Minimum_Depth_of_Binary_Tree_Easy.cpp
/**
155
Minimum Depth of Binary Tree*/


/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: An integer
     */
    int ans;
    int solve_dp(TreeNode *root) {
        if(root == NULL)
            return 0;

        if (root->left == NULL && root->right == NULL)
            return 1;

        int lf = 0x7fffffff, rt = 0x7fffffff;
        if(root->left)
            lf = solve_dp(root->left);

        if(root->right)
            rt = solve_dp(root->right);

        return min(lf, rt) + 1;
    }
    int minDepth(TreeNode *root) {
        // write your code here
        if (!root) return 0;
        return solve_dp(root);
    }
};